src/irace/__init__.py

```diff
@@ -77,29 +74,66 @@ def r_to_python(data):
-            (k,v) for k,v in py_experiment['configuration'].items() if not pd.isna(v)
-        )
+    def parallel_runner(*args, **kwargs):
         try:
-            ret = py_target_runner(py_experiment, py_scenario)
+            experiments = list(r_to_python(args[0]).values())
+            n = len(experiments)
+
+            ans = [None for i in range(n)]
+            for i, experiment in enumerate(experiments):
+                # FIXME: How to skip this conversion?
+                experiment['configuration'] = experiment['configuration'].to_dict('records')[0]
+                # FIXME: We should also filter 'switches'
+                # Filter all the NaN from keys in the dictionary
+                experiment['configuration'] = OrderedDict(
+                    (k,v) for k,v in experiment['configuration'].items() if not pd.isna(v)
+                )
+                if has_worker:
+                    aq.put((i, experiment, scenario_a[0]))
+                else:
+                    res = run_with_catch(target_runner, (experiment, scenario_a[0]), {})
+                    res = check_output_target_runner(ListVector(res), scenario_a[1])
+                    ans[i] = res
+
+            if has_worker:
+                for _ in range(n):
+                    i, res = rq.get()
+                    with localconverter(irace_converter_hack):
+                        res = check_output_target_runner(ListVector(res), scenario_a[1])
+                    ans[i] = res
+
+            return ListVector(zip(range(len(ans)), ans)) 
         except:
+            # rpy2 swallows traceback from any r.rternalize function so we print it manually.
             traceback.print_exc()
-            ret = dict(error=traceback.format_exc())
-        return ListVector(ret)
-    return tmp_r_target_runner
-
-def check_windows(scenario):
-    if scenario.get('parallel', 1) != 1 and os.name == 'nt':
-        raise NotImplementedError('Parallel running on windows is not supported yet. Follow https://github.com/auto-optimization/iracepy/issues/16 for updates. Alternatively, use Linux or MacOS or the irace R package directly.')
+            raise
+    return parallel_runner
+
+def runner_worker(target_runner, aq: Queue, rq: Queue):
+    while True:
+        i, experiment, scenario = aq.get()
+        if i == -1:
+            break
+        rq.put((i, run_with_catch(target_runner, (experiment, scenario), {})))
+
+def check_unsupported_scenarios(scenario):
+    if scenario.get('targetRunnerRetries', 1) > 1:
+        raise NotImplementedError("targetRunnerRetries is not yet supported by the python binding although it's supported in the irace R package. We recommend you to implement retries in your target runner.")
+    if 'targetRunnerParallel' in scenario:
+        raise NotImplementedError("targetRunnerParallel is not yet supported. If you need this feature, consider opening an issue to show us some people actually want to use this.")
+
+def run_irace(irace, args, q: Queue):
```

Comment:
What is this function doing? It does not seem to be used anywhere.

Reply:
Sorry forgot to delete this.